### PR TITLE
ARROW-3674: [Go] Implement Date32 and Date64 array types

### PR DESCRIPTION
**Pull request by dustmop**
Implement both Date32 and Date64 types for arrays. Also resolves ARROW-3675. Unit tests follow the same pattern as the existing float64 and Time{32,64} tests.

### CODE REVIEW

**Review by sbinet — APPROVED** [review]
LGTM

**Comment by dustmop** [comment]
Thanks @sbinet for taking a look, and linking the other JIRA ticket. Given that I don't have write access, is there anything else I need to do to get this merged?

**Comment by sbinet** [comment]
I'd like to wait for another pair of eyes :)
@stuartcarnie and/or @alexandreyc for example.

**Comment by dustmop** [comment]
Sounds good, thanks!

File: go/arrow/numeric.tmpldata
Comment (fsaintjacques) [inline go/arrow/numeric.tmpldata]: Ditto.

File: go/arrow/numeric.tmpldata
Comment (fsaintjacques) [inline go/arrow/numeric.tmpldata]: AFAIK, they're not parametric.

Comment (dustmop) [reply go/arrow/numeric.tmpldata]: Makes sense, there's only a single Date32 unit (same as Date64).

However, when I remove this, the generated code in `numericbuilder.gen.go` uses `arrow.PrimitiveTypes.Date32` within the function `Date32Builder.newData`, but Date32 isn't added to PrimitiveTypes in `datatype_numeric.gen.go`. Is this a bug in the generator or is there something else I need to add? Are both "QualifiedType" and "InternalType" meant to stay?

**Comment by wesm** [comment]
No need to squash

**Comment by wesm** [comment]
@sbinet or @stuartcarnie any final words about this one? I'll merge in a day or two if I don't hear back

**Review by sbinet — APPROVED** [review]
LGTM

**Comment by stuartcarnie** [comment]
Looks great to me, too 👌

**Pull request closed without merging**
